packages/template-block/src/TemplateBlock.tsx

```diff
@@ -0,0 +1,235 @@
+            setLastErrorMessage(error as string);
+        }
+
+        closeTemplateChooser();
+    };
+
+    const handleOpenTemplateChooser = () => openTemplateChooser(onTemplateSelected);
+
+    return (
+        <div data-test-id="template-block-container" className="template-block">
+            {selectedTemplate || isEditing ? (
+                <div
+                    data-test-id="template-block"
+                    className="tw-border tw-border-black-20"
+                    style={{
+                        ...(hasBackground && {
+                            ...getBackgroundColorStyles(backgroundColor),
+                        }),
+                        borderRadius,
+                        border,
+                        padding: hasCustomPaddingValue_blockCard
+                            ? paddingValue_blockCard
+                            : paddingStyleMap[paddingChoice_blockCard],
+                    }}
+                >
+                    {isEditing && lastErrorMessage !== '' && <AlertError errorMessage={lastErrorMessage} />}
+                    <div
+                        data-test-id="template-block-content"
+                        className="tw-flex justify-test"
+                        style={{
+                            flexDirection,
+                            gap: GAP,
+                            alignItems: isRows ? textAnchoringHorizontal : textAnchoringVertical,
+                        }}
+                    >
+                        {hasPreview && (
+                            <TemplatePreview
+                                appBridge={appBridge}
+                                blockSettings={blockSettings}
+                                template={selectedTemplate}
+                                onOpenTemplateChooser={handleOpenTemplateChooser}
+                            />
+                        )}
+                        <div
+                            className={merge(['tw-flex', isRows && hasPreview ? 'tw-flex-col' : 'tw-flex-row'])}
+                            style={{
+                                width: isRows && hasPreview ? `${textRatio}%` : '100%',
+                                gap: GAP,
+                            }}
+                        >
+                            <div className="tw-grow tw-min-w-0">
+                                {
+                                    <TemplateText
+                                        appBridge={appBridge}
+                                        title={templateTitle}
+                                        blockSettings={blockSettings}
+                                        description={templateDescription}
+                                        pageCount={selectedTemplate?.pages.length ?? 0}
+                                        isEditing={isEditing}
+                                        setTitle={(newValue, prevValue) =>
+                                            dispatch({
```

Comment:
You could create a function for this and do the dispatch and the `updateBlockSetings` call in one function.

That way you can get rid of the useEffects that update the block settings.

Generally aim to use as little useEffects as possible.

([Interesting article about the topic](https://react.dev/learn/you-might-not-need-an-effect))

Reply:
Thanks for a hint and really nice article about `useEffect`! 
Specifically, in this case I have to update block settings with a result of data updated inside the reducer. 
I know, that the logic inside reducer is not a really good thing but we have a special case when user chooses the template we have to decide what values should be used for a template's title and description:
if user filled in the custom title and description values before the template was selected then we have to keep that custom values. In other case we're good to use template's title and description though. 
So that's why I need some additional logic inside the reducer and can't update the block settings directly.